File: androidApp/src/main/java/com/konradjurkowski/rickandmorty/android/presentation/character_detail/CharacterDetailViewModel.kt
package com.konradjurkowski.rickandmorty.android.presentation.character_detail

import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.konradjurkowski.rickandmorty.android.presentation.navigation.NavigationArguments.CHARACTER_ID
import com.konradjurkowski.rickandmorty.android.presentation.navigation.NavigationArguments.CHARACTER_NAME
import com.konradjurkowski.rickandmorty.interactors.character_detail.GetCharacter
import com.konradjurkowski.rickandmorty.presentation.character_detail.CharacterDetailEvent
import com.konradjurkowski.rickandmorty.presentation.character_detail.CharacterDetailState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import javax.inject.Inject

@HiltViewModel
class CharacterDetailViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    private val getCharacter: GetCharacter
): ViewModel() {

    var state by mutableStateOf(CharacterDetailState())

    init {
        savedStateHandle.get<Int>(CHARACTER_ID)?.let { characterId ->
            onTriggerEvent(CharacterDetailEvent.GetCharacter(characterId))
        }
        savedStateHandle.get<String>(CHARACTER_NAME)?.let { characterName ->
            state = state.copy(pageTitle = characterName)
        }
    }

    private fun onTriggerEvent(event: CharacterDetailEvent) {
        when(event) {
            is CharacterDetailEvent.GetCharacter -> {
                loadCharacter(event.characterId)
            }
        }
    }

    private fun loadCharacter(characterId: Int) {
        getCharacter
            .execute(characterId)
            .onEach { dataState ->
                state = state.copy(isLoading = dataState.isLoading)

                dataState.data?.let { character ->
                    state = state.copy(character = character)
                }

                dataState.message?.let { message ->
                    state = state.copy(error = message)
                }
            }.launchIn(viewModelScope)
    }
}

File: iosApp/iosApp/presentation/character_detail/CharacterDetailViewModel.swift
//
//  CharacterDetailViewModel.swift
//  iosApp
//
//  Created by Konrad Jurkowski on 08/11/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import shared

class CharacterDetailViewModel: ObservableObject {
    
    // dependencies
    private let getCharacter: GetCharacter
    
    @Published var state: CharacterDetailState = CharacterDetailState()
    
    @Published var showDialog: Bool = false
    
    init(
        charactrId: Int32,
        getCharacter: GetCharacter
    ) {
        self.getCharacter = getCharacter
        onTriggerEvent(event: CharacterDetailEvent.GetCharacter(characterId: charactrId))
    }
    
    func onTriggerEvent(event: CharacterDetailEvent) {
        switch event {
            case is CharacterDetailEvent.GetCharacter:
            loadCharacter(characterId: (event as! CharacterDetailEvent.GetCharacter).characterId)
            default: ()
        }
    }
    
    func removeErrorMessage() {
        self.updateState(error: nil)
    }
    
    private func loadCharacter(characterId: Int32) {
        getCharacter
            .execute(characterId: characterId)
            .watch(
                coroutineScope: nil,
                callback: { dataState in
                    if dataState != nil {
                        let data = dataState?.data
                        let message = dataState?.message
                        let loading = dataState?.isLoading
                        
                        self.updateState(isLoading: loading)
                        
                        if data != nil {
                            self.updateState(character: data! as Character)
                        }
                        
                        if message != nil {
                            self.updateState(error: message)
                        }
                    } else {
                        self.updateState(error: "Cannot load character!")
                    }
                })
    }
    
    private func updateState(
        isLoading: Bool? = nil,
        character: Character? = nil,
        error: String? = nil
    ) {
        state = state.doCopy(
            isLoading: isLoading ?? state.isLoading,
            character: character ?? state.character,
            pageTitle: "",
            error: error ?? state.error
        )
        shouldShowDialog()
    }
    
    private func shouldShowDialog() {
        showDialog = state.error != nil
    }
}
